### ISSUE TITLE
[BE] [Impl] Connecting MongoDB and Elasticsearch

### ISSUE THREAD

**Issue by Leats** [Drakulix/knex#87]
### Research : Syncing MongoDB and Elasticsearch
I've looked into tools for syncing, and these were the mentioned options:
 
#### [mongo-connector](https://pypi.python.org/pypi/mongo-connector) 
creates a pipeline from a MongoDB cluster to one or more target systems, like Elasticsearch. It synchronizes data in MongoDB to the target then tails the MongoDB oplog, keeping up with operations in MongoDB in real-time. Can be easily installed with pip and seems to be the popular option since mongodb river is deprecated (for river read further). 
See also the wiki of mongo-connector: [Usage with Elasticsearch](https://github.com/mongodb-labs/mongo-connector/wiki/Usage-with-Elasticsearch)
 
#### [Transporter](https://github.com/compose/transporter) 
is for synchronizing data e.g. from mongoDB to elasticsearch, however, that does come as a one time export and would need to be modified to keep it in sync (feature only in beta).
 
#### [logstash](https://www.elastic.co/guide/en/logstash/current/introduction.html)
Logstash is an open source data collection engine with real-time pipelining capabilities. Logstash can dynamically unify data from disparate sources and normalize the data into destinations of your choice.
Has a plugin for output to elasticsearch, but only a community-maintained non-verified plugin for mongodb input. Probably possible but doesn’t seem like the easiest option.
 
[MongoDB river](https://github.com/richardwilly98/elasticsearch-river-mongodb) seems to have been a popular choice, as well. But that one is deprecated and won’t work with the newest MongoDB version.
 
### Conclusion
In the end I’d personally stick to mongo-connector, which had been suggested at first, too. Easily set up and actually made exactly for our purpose.

**Comment by Drakulix** [comment]
Very nice write up.

How difficult is excluding/black-listing or white-listing certain collections (we basically only want to index our project collection from mongodb)? This would also be very interesting.

Other then that everything considered, I would probably select mongo-connector.

**Comment by Leats** [comment]
From the [Configuration Options](https://github.com/mongodb-labs/mongo-connector/wiki/Configuration-Options) of mongo-connector, these seem to be the possible black-listing/white-listing options:
You can configure **namespaces** to include or [exclude](https://github.com/mongodb-labs/mongo-connector/wiki/Configuration-Options#excluding-namespaces) certain ones
And you can configure which **fields** to [exclude](https://github.com/mongodb-labs/mongo-connector/wiki/Configuration-Options#exclude_fields) or include from documents

**Comment by Drakulix** [comment]
Promoted for implementation.
Setup Mongo-Connector as docker container for mongodb -> elastic search

**Comment by Leats** [comment]
Added new container on branch [backend/feature/mongo-connector/iss87](https://github.com/Drakulix/knex/tree/backend/feature/mongo-connector/iss87), though so far it's mostly testing if it's alright like this. 
I also added a config json for it where one can blacklist/whitelist the needed namespaces.

However, so far mongo-connector exits with "pymongo.errors.ServerSelectionTimeoutError: localhost:27017: [Errno 111] Connection refused", which probably means there need to be modifications made to mongodb (Replica Set of mongoDB as stated in the [mongo-connector guide](https://github.com/mongodb-labs/mongo-connector/wiki/Getting-Started)?). I haven't figured out how to do that properly yet, but wanted to give info about the WIP.

**Comment by Drakulix** [comment]
@Leats you should not use localhost. The other containers are actually seen as different machines by your new container. That means you should be using something like `mongodb:27017`. The names correspond to the container names in the docker-compose.yml.

**Comment by Leats** [comment]
Oh, that does make more sense, thank you! Updated that and looked into another issue with it after changing the config. It does indeed need a replica set of docker, which I don't know how to do yet and hope won't make any problems with the already existing code.
Will look into that until tomorrow.

**Comment by Leats** [comment]
Okay, I'm so far that apparently mongodb recognizes that I want to create a replica set but the only other info I have is that mongo-connector can't find it. Then again I'm not sure if the replica set is ever actually initialized properly...

`mongodb_1          | 2017-06-12T13:06:22.497+0000 I CONTROL  [initandlisten] options: { replication: { replSet: "rs0" } }`

`mongo-connector_1  | 2017-06-12 13:06:28,131 [ERROR] mongo_connector.connector:373 - No replica set at "mongodb:27017"! A replica set is required to run mongo-connector. Shutting down...`

I don't see any other errors in the console from mongodb or mongo-connector. I am actually really not sure how to proceed? I mean, it all seems really simple but I think I'm not really understanding how to do it properly. 
Maybe there has to be a different approach.

**Comment by Drakulix** [comment]
Maybe mongo-connector needs the name of the replSet?

**Comment by Leats** [comment]
Problem was, that mongodb wasn't really starting to work as a replica set, which is why mongo-connector obviously didn't find it. That works now, for most of the time(?).

And honestly, I'm sorry but I have to give up. I'm sure someone else actually would be better fit to do this, I'm sure I did some really stupid mistakes or something. And I don't want to waste any more time for the project until I _might_ get to the solution.

I don't even know anything anymore, first I thought it was doing well, then one time mongo-connector gets the connection refused by elasticsearch (that's happening most of the time), sometimes mongodb doesn't manage to form a Primary replica set member (just rarely), and sometimes nothing happens at all regarding the connector (and I hadn't even changed the code).

I updated the branch I linked above to what the last status is.

I mean, the one thing which should work is installing mongo-connector. Just connecting with it is... nothing I managed.

Edit: if anyone has tips or ideas i of course will try again, but for now i really don't know anympre :(
